proposals/cmplx-real-pointers.txt

```diff
@@ -0,0 +1,139 @@
+  compilers store a complex number by the sequence 
+  real part/imaginary part.
+
+For instance, consider a code that performs a FFT based convolution
+at some point:
+
+    program foo
+    real, allocatable :: r1(:), r2(:), r3(:)
+    integer, parameter :: n = 100000
+    ! ...
+    allocate( r1(n), r2(n), r3(n) )
+    ! ... some computations on r1(:) and r2(:) as reals
+    call rfft(r1)
+    call rfft(r2)
+    call fftconvol(r1,r2,r3,n/2)   ! called without any interface
+    call irfft(r3)
+    ! ... some computations on r3(:) as real
+    end program
+
+    subroutine fftconvol(c1,c2,c3,n)
+    integer, intent(in) :: n
+    complex, intent(in)  :: c1(n), c2(n)
+    complex, intent(out) :: c3(n)
+    c3(:) = c1(:) * c2(:)
+    end subroutine
+
+	
+3. Proposed solution
+
+The proposal essentially aims at standardizing the usual practice 
+above, by allowing a real array to be viewed as a complex array and 
+vice-versa. The above code would become:
+
+    program foo
+    real, allocatable, target :: r1(:), r2(:), r3(:)
+    integer, parameter :: n = 100000
+    complex, pointer :: c1(:), c2(:), c3(:)
+    ! ...
+    allocate( r1(n), r2(n), r3(n) )
+    ! ... some computations on r1(:) and r2(:) as reals
+    call rfft(r1)
+    call rfft(r2)
+    c1 => complex_pointer(r1)
+    c2 => complex_pointer(r2)
+    c3 => complex_pointer(r3)
+    c3(:) = c1(:) * c2(:)
+    call irfft(r3)
+    ! ... some computations on r3(:) as real
+    end program
+
+3.1. New intrinsic functions
+
+`c => complex_pointer(r)`
+
+- `r` is a *contiguous* real array, which has either the target or the
+  pointer attribute
+- `c` is a complex array pointer of the same rank than `r`
+  - `c` can also be a complex scalar pointer, in the case where r is a
+    rank-1 array of size 2
+- the size of the first dimension of `r` shall be even
+- `c%re` refers to the same storage than `r(1::2)` (rank-1), or 
+  `r(1::2,:)` (rank-2), etc...
+- `c%im` refers to the same storage than `r(2::2)` (rank-1), or 
+  `r(2::2,:)` (rank-2), etc...
+
+`r => real_pointer(c)`
+
+- the exact opposite
+- `c` is a *contiguous* complex array or a complex scalar, which has 
+  either the target or the pointer attribute
+- `r` is a complex array pointer of the same rank than `c`
+  - if `c` is a scalar, then `r` is a rank-1 pointer of size 2
+- same other rules as above
+
+3.2 Alternative syntax
+
+Alternative syntax 1 (similar to the c_f_pointer() subroutine):
+```
+call complex_pointer(r,c[,shape])
+call real_pointer(c,r[,shape])
+```
+
+Alternative syntax 2 (kind of pointer casting):
+```
+c => complex :: r
+r => real :: c
+```
+
+4. Objections
+
+4.1. Memory layout
+
+This proposal would constraint the memory layout of the complex type 
+beyond what the current standard does. However, the required layout 
+seems to be the de-facto standard, used by virtually all existing 
+compilers, and it's hard to think about a possible drawback of 
+constraining the layout like this.
```

Comment:
The drawback of constraining the layout is that we can't represent the "split" (block) layout natively, but the split layout might have better performance. (We can't do the split layout today, but this might be a good extension in the future.)

Here is a page in FFTW that talks about this: https://fftw.org/fftw3_doc/Interleaved-and-split-arrays.html.

Here is another: https://developer.apple.com/documentation/accelerate/data_packing_for_fourier_transforms?language=objc

It looks like the terms for this are:

* Interleaved-Complex Format
* Split-Complex Format

The advantage of the spit-complex format is performance (and Fortran cares about performance). Here is a page that talks about it: https://www.jobilize.com/course/section/split-format-vs-interleaved-format-by-openstax, citing:

> The benefit of operating in split format is obvious: the shuffle operations that were required in [[link]](https://www.jobilize.com/online/course/0-2-implementation-details-computing-the-fast-fourier-transform-by-ope?page=6#uid20) are avoided because the real and imaginary parts can be implicitly swapped at the instruction level, rather than by awkwardly manipulating SIMD registers at the data level of abstraction.

Roughly speaking, the interleaved-complex format requires a shuffle (swap) SIMD instruction for many complex operations (such as multiply) that the split-complex format doesn't.


Coming to your proposal: one way to resolve it is that for interleaved-complex format you would do:
```
c1 => complex_pointer(r1)
```
And for split-complex format you would do:
```
c1 => complex_pointer(r1_real, r1_imag)
```
Where the `c1` complex array is declared with the appropriate `layout(interleaved)` and `layout(split)` layouts.

As long as we keep this in mind (to not close the door to the split format), then your proposal should work.

Reply:
As discussed in the issue, I have some serious doubts about the possibility of introducing a complex split layout in the standard, as it would result in scalars that are possibly stored non contiguously. Assuming it's possible, I think that the proposal doesn't really close the door:
> the obtained view would be layout dependent, and the standard would have to describe the view for each layout
In the case of a split layout, `c%re` would refer to the same storage as `r(1:n/2)`

I can make it more explicit in the proposal.

Reply:
Excellent, your commit https://github.com/j3-fortran/fortran_proposals/pull/325/commits/a928885dce400f26f5fbd08c0556ba9868344403 addresses my comment, thank you. Now it's good I think.